src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, can_reshape, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (is_optimized_multiplication()) {
```

Comment:
What would happen if that `if` is removed? Would matching in the rest of the method accept something that's not a valid pattern?

Reply:
It wouldn't accept invalid patterns. It still makes (semantically) correct transformation but there is a risk it can't progress. The `if` here is to safeguard repeatedly adding terms of an already optimized mul node. Otherwise there would be an eventual timeout/live node limit situation. (E.g., `3 * a` => `(a << 2) + a` => `3 * a` => ...)